cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs

```diff
@@ -3181,9 +3184,27 @@ pMaxTransactionSize =
       , Opt.help "Maximum transaction size."
       ]
 
+-- | @wordReader typeName@ is a reader for a word of type @a@. When it fails
+-- parsing, it provides a nice error message. This custom reader is needed
+-- to avoid the overflow issues of 'Opt.auto' described in https://github.com/IntersectMBO/cardano-cli/issues/860.
+wordReader :: (Integral a, Bits a) => String -> ReadM a
+wordReader typeName =
+  Opt.maybeReader parser
+ where
+  -- We use 'readMaybe' instead of 'read', because the latter wouldn't output a nice error message
+  parser s =
+    case readMaybe s >>= naturalToWordMaybe >>= toIntegralSized of
+      Nothing ->
+        fail $ "Cannot parse " <> s <> " as a " <> typeName
+      Just a ->
+        Just a
```

Comment:
what do you think about removing the function argument:
```suggestion
wordReader :: forall a. (Typeable a, Integral a, Bits a) => String -> ReadM a
wordReader =
  Opt.maybeReader parser
 where
  -- We use 'readMaybe' instead of 'read', because the latter wouldn't output a nice error message
  typeName = show $ typeRep (Proxy @a)
  parseFailure = fail $ "Cannot parse " <> s <> " as a " <> typeName
  parser s =
    fromMaybe parseFailure $
      readMaybe s >>= naturalToWordMaybe >>= toIntegralSized
```

Reply:
Oh actually, I've just noticed the type `parser :: String -> Maybe a`. So I think you should use actually https://hackage.haskell.org/package/optparse-applicative-0.18.1.0/docs/src/Options.Applicative.Builder.html#eitherReader and return `Left` instead of `fail`.

Reply:
@carbolymer> done changes:

1. Use `eitherReader`
2. Remove variable for type name